File: projects/bmc_alarm_tool/alarm_operator.py
#!/usr/bin/python

import sys
import subprocess

def run_cmd(cmd):
  '''
  use subprocess.popen() to run commands
  '''
  try:
    p = subprocess.Popen(cmd, shell=True, stdin=subprocess.PIPE, stdout=subprocess.PIPE)
    (result, err) = p.communicate(cmd)
    return (result, err)
  except Exception, ex:
    print 'ERROR: Failed to run command <%s>' %cmd
    raise ex

def display_usage():
  print '\n*=======================================================================================================================*'
  print '\033[7m* Name        : Alarm Operator'
  print '* Description : This tool is used to raise/clear an appointed BMC Alarm from BMC to OSS.'
  print '* Usage       : ./alarm_operator.py <operation> <managed device> <managed object> <alarm ID> <severity> <"alarm text">'
  print '*               allowed values of <operation> : raise / clear'
  print '*               allowed values of <severity>  : INDETERMINATE / CRITICAL / MAJOR / MINOR / WARNING'
  print '* Example     : ./alarm_operator.py raise BMC SDCH 449153262 MAJOR "Setting is missing sa service/class."'
  print '* Note        : The "alarm ID" is defined in file "alarm_definition.txt".\033[0m'
  print '*=======================================================================================================================*\n'

def validate_mode(mode):
  if not mode in ('raise', 'clear'):
    return False
  return True

def validate_severity(severity):
  if not severity in ('INDETERMINATE', 'CRITICAL', 'MAJOR', 'MINOR', 'WARNING'):
    return False
  return True

def operate_alarm(mode, managedDevice, managedObj, alarmId, severity, alarmText):
  #cmd = 'java -cp .:/home/led/work/temp/bmc_alarm_tool/lib/* -jar AlarmOperator.jar %s %s %s %s %s "%s" 2>&1' %(mode, managedDevice, managedObj, alarmId, severity, alarmText)
  cmd = 'java -jar AlarmOperator.jar %s %s %s %s %s "%s" 2>&1' %(mode, managedDevice, managedObj, alarmId, severity, alarmText)
  (result, err) = run_cmd(cmd)
  lines = result.split('\n')
  for line in lines:
    if not 'log4j:' in line:
      print line
      
if __name__ == '__main__':
  if len(sys.argv) != 7:
    display_usage()
    #print 'ERROR: parameter count is wrong:', len(sys.argv)
    exit(1)

  mode = sys.argv[1].lower()
  if not validate_mode(mode):
    #display_usage()
    print 'ERROR: operation is wrong:', sys.argv[1]
    exit(1)
    
  managedDevice = sys.argv[2]
  managedObj = sys.argv[3]
  alarmId = sys.argv[4]
  severity = sys.argv[5].upper()
  if not validate_severity(severity):
    #display_usage()
    print 'ERROR: severity is wrong:', sys.argv[5]
    exit(1)
    
  alarmText = sys.argv[6]
  
  operate_alarm(mode, managedDevice, managedObj, alarmId, severity, alarmText)
  
  

  

File: projects/bmc_alarm_tool/AlarmOperator.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.net.Socket;
import java.net.InetAddress;
import java.util.Map;
import java.util.HashMap;

import com.tailf.cdb.Cdb;
import com.tailf.conf.ConfBuf;
import com.tailf.conf.ConfDatetime;
import com.tailf.conf.ConfException;
import com.tailf.conf.ConfIdentityRef;
import com.tailf.conf.ConfKey;
import com.tailf.conf.ConfPath;
import com.tailf.maapi.Maapi;
import com.tailf.maapi.MaapiUserSessionFlag;
import com.tailf.ncs.ApplicationComponent;
import com.tailf.ncs.NcsMain;
import com.tailf.ncs.alarmman.common.Alarm;
import com.tailf.ncs.alarmman.common.AlarmId;
import com.tailf.ncs.alarmman.common.ManagedDevice;
import com.tailf.ncs.alarmman.common.ManagedObject;
import com.tailf.ncs.alarmman.common.PerceivedSeverity;
import com.tailf.ncs.alarmman.producer.AlarmSink;

public class AlarmOperator {
    private static final int hash = 959898198;
    private static final String RAISE = "raise";
    private static final String CLEAR = "clear";

    private static int mode = 0;
    private static String managedDevice;
    private static String managedObj;
    private static int alarmId;
    private static PerceivedSeverity severity = PerceivedSeverity.MAJOR;
    private static String alarmText;
    private static Map alarmMap = new HashMap();
    
    public static void main(String[] args) {
        if (!initAlarmList()) {
            System.exit(1);
        }
        
        if (!parseParams(args)) {
            System.exit(1);
        }

        switch (mode) {
            case 1:
                raiseAlarm();
                break;
            case 2:
                clearAlarm();
                break;
            default:
                System.out.println("ERROR: The alarm operation is invalid.");
                break;
        }
    }

    private static boolean initAlarmList() {
        String alarmFile = "./alarm_definition.txt";
        BufferedReader br = null;
        try {
            br = new BufferedReader(new FileReader(alarmFile));
            StringBuilder sb = new StringBuilder();
            String line = br.readLine();

            while (line != null) {
                String[] subLines = line.split("=");
                alarmMap.put(Integer.parseInt(subLines[1]), subLines[0]);
                line = br.readLine();
            }
            br.close();
            System.out.println("Load and initialize alarm list successfully.");
            return true;
        } catch (Exception e) {
            try {
                br.close();
            } catch (IOException e1) {
                System.out.println("ERROR: Failed to close file.");
            }
            System.out.println("ERROR: Failed to load and initialize alarm list : " + e);
            return false;
        } 
    }
    
    private static boolean parseParams(String[] args) {
        /*if (args.length != 6) {
            //displayUsage();
            System.out.println("ERROR: not enough args: " + args.length);
            return false;
        }*/
        
        if (args[0].equals(RAISE)) {
            mode = 1;
            severity = PerceivedSeverity.valueOf(args[4]);            
        } else if (args[0].equals(CLEAR)) {
            mode = 2;
        } else {        	  
            //displayUsage();
            return false;
        }
        
        managedDevice = args[1];
        managedObj = args[2];
        try {
          	alarmId = Integer.parseInt(args[3]);
          	if (!alarmMap.containsKey(alarmId)) {
            		System.out.println("ERROR: The alarm ID is invalid : " + alarmId);
            		//displayUsage();
            		return false;
          	}
        } catch(Exception e) {
        	  System.out.println("ERROR: The alarm ID is invalid : " + alarmId);
    		    //displayUsage();
        	  return false;
        }        
        
        alarmText = args[5];
        
        return true;
    } 

    private static void displayUsage() {
        System.out.println("Usage: ./alarmOperator.sh <raise/clear> <managed device> <managed object> <alarm ID> <severity> <\"alarm text\">");
        System.out.println("Example: ./alarmOperator.sh raise BMC SDCH MAJOR 449153262 \"Setting is missing sa service/class.\"");
    }
    
    private static void raiseAlarm() {
        try {
            if (!operateAlarm(severity)) {
                System.out.println("ERROR: Failed to raise alarm");
            }
        } catch (Exception e) {
            System.out.println("ERROR: Failed to raise alarm:" + e);
        }
    }
    
    private static void clearAlarm() {
        try {
            if (!operateAlarm(PerceivedSeverity.CLEARED)) {
                System.out.println("ERROR: Failed to clear alarm");
            }
        } catch (Exception e) {
            System.out.println("ERROR: Failed to clear alarm:" + e);
        }
    }
    
    private static boolean operateAlarm(PerceivedSeverity svt) throws Exception {
        Socket socket = null;
        System.out.println("Start sending alarm: ID<" + alarmId + ">, Type<" +
                alarmMap.get(alarmId) + ">");
        try {
            socket = new Socket(NcsMain.getInstance().getNcsHost(), 
                NcsMain.getInstance().getNcsPort() );
            Cdb cdb = new Cdb(AlarmOperator.class.getName(), socket);
            AlarmSink alarmSink = new AlarmSink(cdb);
            Alarm alarm = new Alarm(new ManagedDevice(managedDevice), 
                    new ManagedObject(managedObj),
                    new ConfIdentityRef(hash, alarmId), 
                    svt,
                    false, alarmText, 
                    new ArrayList<ManagedObject>(), 
                    new ArrayList<AlarmId>(), 
                    new ArrayList<ManagedObject>(),
                    ConfDatetime.getConfDatetime());
                    
            alarmSink.submitAlarm(alarm);
            
            socket.close();
            System.out.println("Send alarm successfully");
            return true;
        } catch (Exception e1) {
            try {
                socket.close();
            } catch (Exception e) {
                System.out.println("ERROR: close socket failed: " + e);
            }
            System.out.println("ERROR: Failed to operate alarm: " + e1);
            return false;
        }
    }    
}
